### PR TITLE
Update backend.Key to be its own distinct type

### LINKED ISSUE #6088: Not able to delete remote clusters that contain "....." in the name
### Description

**What happened**:

When customer runs `tctl get rc`, they notice there are entries like `cl2-............-c3c2fb` Running `tctl rm rc/cl2-............-c3c2fb` results in error message:

```error: invalid path, path can only be composed of characters, hyphens, slashes, and dots```

**What you expected to happen**:

Customer should be able to delete nodes/clusters....etc., with trailing periods in the name.

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1. Create leaf cluster with multiple sequential periods in the name.
2. Attempt to remove the cluster.
3. Observe the process failing due to aforementioned error.

### Server Details
- Teleport version (run `teleport version`): 4.4+

### LINKED ISSUE #9107: Remove ability to add / in connector definitions
It is possible to accidentally insert a forward slash / when adding a SAML connector. The forward slash is not rejected as a reserved character (as with underscores) and the connector is saved.

This connector is then presented to end users on the WebUI page.

It is not possible to remove the connector using `tctl` - the command interprets the slash as part of the `saml/` contruct required to get/ delete the connector.

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1. Create a connector with the name `this/test` of type SAML
2. Attempt to delete the connector using the UI:. Error is reported:
```
404 - https://example.teleport.com:3080/v1/enterprise/saml/this%2Ftest
```

3. Attempt to delete the connector using tctl, also with escaping:

```
root@ubuntu1:~#  tctl rm saml/this/test

ERROR REPORT:
Original Error: *trace.NotFoundError key /web/connectors/saml/connectors/test is not found
Stack Trace:
	/go/src/github.com/gravitational/teleport/vendor/github.com/gravitational/teleport/api/client/client.go:1173 github.com/gravitational/teleport/api/client.(*Client).DeleteSAMLConnector

root@ubuntu1:~# tctl rm saml\/this\/test

ERROR REPORT:
Original Error: *trace.NotFoundError key /web/connectors/saml/connectors/test is not found
Stack Trace:
	/go/src/github.com/gravitational/teleport/vendor/github.com/gravitational/teleport/api/client/client.go:1173 github.com/gravitational/teleport/api/client.(*Client).DeleteSAMLConnector
```
4. Note that the connector remains visible via `tctl get connectors`

We can see the erroneous entry in the metadata in back-end DynamoDB - is it possible to remove the connector by deleting this entry without compromising the cluster.

### Server Details
- Teleport version (run `teleport version`): 7.3.0
- Server OS (e.g. from `/etc/os-release`): Ubuntu 21.04
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): Local VM, AWS
- Additional details:

### Client Details
- Tsh version (`tsh version`): 7.3.0
- Computer OS (e.g. Linux, macOS, Windows): Linux/ macOS
- Browser version (for UI-related issues): Brave/ Chrome
- Installed via (e.g. apt, yum, brew, website download): website download
- Additional details:

### Debug Logs
Please include or attach debug logs, when appropriate. Obfuscate sensitive information!
- Start Teleport with --debug flag (`teleport --debug`)
- Run tsh with --debug flag (`tsh --debug`)

### LINKED ISSUE #10576: Creating the user with / in name is possible and breaks things
### What

Found during v9 beta tests.

Both Web UI and CLI allow '/' to be within the name of the user created. The user is created in backend (verified by raw db query) and the creation shows up in audit log, but things break in weird ways with such a user:
- when setting the password via invite link an error is thrown after "create account" is clicked. It reads "object not found" which is not very helpful.
- `tctl get users` does not list such a user
- probably other problems too...

### Potential solution

1. Disallow creation of such users, or:
2. Make sure they work as intended.

### Other problems

This may well affect other objects that we create due to the way backend addressing works.

### LINKED ISSUE #42823: Role created with / in name cannot be deleted
Expected behavior:
* role can be deleted with `tctl edit` or `tctl rm`

Current behavior:
```shell
$ tctl edit role/a/b
ERROR: could not get resource role/a/b: role b is not found
        role b is not found
$ tctl rm role/a/b
ERROR: role "b" is not found
```

Same happens in the WebUI

Bug details:
- Teleport version: 15.4.0 community
- Recreation steps: create user with `/` in name, then try delete in the webUI or with `tctl`
- Debug logs: none applicable

### PR DESCRIPTION
**Pull request by rosstimothy**
Converts the `backend.Key` from a slice of bytes to a concrete struct. The main motivation behind this change is to be able to better distinguish individual components of a key. The textual representation of a backend key is constructed by joining all components of the key with a `/`. By representing the entire key as a single textual object it prevented resources containing a `/` in their name from being properly identified. There was a lot of code that interpolated keys in various manners and expected only a specific number of subcomponents for a particular resource. This lead to most, if not all, of the RPCs used to create a resource to permit a name containing `/`, (i.e. a user named `test/llama/1`), but any RPCs used to retrieve the resources would fail to retrieve them from the backend. 

Every `backend.Key` now carries a slice of all the individual components in addition to the textual representation. This permits more fine grained inspection per component, while also not having to construct the textual representation for a group of components more than once.

The `backend.Sanitizer` was updated to only validate keys for backend writes. Retrieving and deleting invalid of malformed keys is however permitted. This will allow any existing resources that were created with `/` in one of the components to become visible in the UI or tctl get and deletable. However, any new resources with ` /` in there name are explicitly prevented.

Closes #6088 
Closes #9107
Closes #10576
Closes #42823

### CODE REVIEW

File: lib/backend/key.go [hunk 1]
```diff
@@ -19,81 +19,138 @@ package backend
...
 const Separator = '/'
+const SeparatorString = string(Separator)
 
 // NewKey joins parts into path separated by Separator,
 // makes sure path always starts with Separator ("/")
-func NewKey(parts ...string) Key {
-	return internalKey("", parts...)
+func NewKey(components ...string) Key {
+	k := internalKey("", components...)
+	k.exactKey = k.s == SeparatorString || (len(k.s) > 0 && k.s[len(k.s)-1] == Separator)
```
Comment (espadolini) [inline lib/backend/key.go:42]:
```suggestion
	k.exactKey = len(k.s) > 0 && k.s[len(k.s)-1] == Separator
```

Isn't `internalKey` always returning a nonempty key, anyway?

File: lib/backend/key.go [hunk 2]
```diff
@@ -19,81 +19,138 @@ package backend
...
+		exactKey:   s == SeparatorString || (len(s) > 0 && s[len(s)-1] == Separator),
+		noEnd:      s == string(noEnd),
+	}
+}
+
+func (k Key) IsZero() bool {
+	return len(k.components) == 0 && k.s == ""
+}
+
+func internalKey(internalPrefix string, components ...string) Key {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/backend/key.go:74]: +1, it seems like that should be the default behavior, and silently splitting components should be some kind of opt-in for back-combat.

Comment (fspmarshall) [inline lib/backend/key.go:74]: This function is pretty dense.  I think some detailed comments would be helpful.

**Review by fspmarshall — APPROVED** [review]
LGTM once existing feedback is addressed.

File: lib/services/local/trust_test.go [hunk 3]
```diff
@@ -132,7 +132,7 @@ func TestRemoteClusterCRUD(t *testing.T) {
 	// make sure it's really gone
 	err = trustService.DeleteRemoteCluster(ctx, "foo")
 	require.Error(t, err)
-	require.ErrorIs(t, err, trace.NotFound("key \"/remoteClusters/foo\" is not found"))
+	require.ErrorIs(t, err, trace.NotFound("key /remoteClusters/foo is not found"))
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/services/local/trust_test.go:135]: It's probably fine now that we're better at enforcing key sanitization? 🤷

**Review by espadolini — APPROVED** [review]

**Review by espadolini — APPROVED** [review]

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**